### ISSUE TITLE
[BUG][Fuzzer][wrong-code] `_tir_u32_to_int_to_float` silently decodes signed sub-word ints as unsigned (zero-extends instead of sign-extending) — every negative value decodes positive

### ISSUE THREAD

**Issue by Hughshine** [tile-ai/tilelang#2482]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### What version of TileLang are you using?

0.1.11

### System information

- Installed from the `0.1.11` wheel; source pinned to tag `v0.1.11` (commit `cd37ed5fc35ae7a60a1277c8eb49028174ac51e6`).
- Python 3.13, linux
- GPU: **NVIDIA L40S** (sm_89). The defect is in a `tilelang.quantize` TIR convert; it is host/arch-independent.

### Problem description

`_tir_u32_to_int_to_float` is named for decoding a **signed** sub-word integer (`..._to_int_to_float`) but its body only masks the low `nbit` bits — a **zero-extend, with no sign-extension** — so every negative value decodes to a positive one (a stored signed int4 `-8`, nibble `0x8`, decodes to `+8.0`). It compiles and runs with no error: a silent miscompile of signed sub-word dequantization.

It is **exported/user-callable** (`tilelang.quantize.quantization`) with no in-library caller, so the affected user is anyone who calls this convert on signed packed data — but a wrong implementation of an exported, signed-named function is the defect regardless of who reaches it.

<details>
<summary>Root cause — body masks (zero-extends); the sign-extend step is absent</summary>

[`quantization.py#L110-L113`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/tilelang/quantize/quantization.py#L110-L113):

```python
def _tir_u32_to_int_to_float(nbit, val, pos, dtype):
    assert val.dtype == "uint32"
    mask = const((1 << nbit) - 1, "uint32")
    return Cast(dtype, (val >> (pos * nbit)) & mask)   # `& mask` = zero-extend; no sign-extend
```

The `& mask` keeps only the low `nbit` bits, so the top bit of a signed sub-word (its sign) is dropped before the `Cast` — the value is read as unsigned. The implemented, in-use sibling for the same "signed sub-word → to_int" job, [`_tir_packed_int_to_int_to_float` (`quantization.py#L128-L138`)](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/tilelang/quantize/quantization.py#L128-L138), instead does `(unextended << (32 - nbit)) >> (32 - nbit)` — a real arithmetic-shift sign-extension. `_tir_u32_to_int_to_float` is missing exactly that step.
</details>

### Reproducible example code

Note: use a real `.py` file (eager JIT calls `inspect.getsource`; `python -c` trips an unrelated `OSError`). `tilelang.disable_cache()`.

```python
import tilelang, tilelang.language as T, torch
tilelang.disable_cache()
from tilelang.quantize.quantization import _tir_u32_to_int_to_float

NBIT = 4
vals = [-8, -1, 7, -2, 0, -8, 3, -5]            # signed int4
packed = 0
for i, v in enumerate(vals):
    packed |= (v & 0xF) << (i * NBIT)           # pack into one uint32

@T.prim_func
def main(I: T.Tensor((1,), "uint32"), O: T.Tensor((8,), "float32")):
    with T.Kernel(1, threads=1) as _:
        for i in T.serial(8):
            O[i] = _tir_u32_to_int_to_float(NBIT, I[0], i, "float32")

k = tilelang.compile(main, target="cuda")
I = torch.tensor([packed], dtype=torch.uint32, device="cuda")
O = torch.zeros(8, dtype=torch.float32, device="cuda")
k(I, O); torch.cuda.synchronize()
print(O.cpu().tolist())     # got:  [8, 15, 7, 14, 0, 8, 3, 11]   (every negative is positive)
print(vals)                 # want: [-8, -1, 7, -2, 0, -8, 3, -5]
```

The control is the sign-extending sibling on the **identical** packed bits: `_tir_packed_int_to_int_to_float(32)(NBIT, I[0], i, "float32")` decodes to the correct `[-8, -1, 7, -2, 0, -8, 3, -5]`, proving the packing/wiring is sound and the error is solely in this function.

### Traceback

```pytb
No traceback — the kernel compiles and runs to completion; the signed values are silently decoded as unsigned (every negative becomes positive), deterministically.
```

### Expected behavior

`_tir_u32_to_int_to_float`, given its `..._to_int_to_float` (signed) name, should sign-extend each signed sub-word value so a stored `-8` decodes to `-8.0`, **or** be renamed/documented as the unsigned decode it actually is. This is grounded in the project's own code: the in-use sibling `_tir_packed_int_to_int_to_float` performs exactly the `(x << (32-nbit)) >> (32-nbit)` sign-extension for the same signed sub-word job, so the intended behavior of a signed-named convert is sign extension, not zero extension.

### Additional context

**Suggested fix (proposed, not built).** Apply the same arithmetic-shift sign-extension the sibling uses — replace `(val >> (pos*nbit)) & mask` with an `(x << (32-nbit)) >> (32-nbit)` style extension of the `nbit` field (signed shift), so the sign bit is preserved before the `Cast`.

Confirmed on an L40S (sm_89, 0.1.11) by acting as a user — a real dequant kernel packing signed int4 `[-8,-1,7,-2,0,-8,3,-5]` decodes via this function to `[8,15,7,14,0,8,3,11]` (every negative wrong; `-8 → +8.0`), while the sign-extending sibling on the identical bits gives the correct values. I did not build the fix.

**Provenance.** The function shipped with the zero-extend from the start: it was introduced by [#494](https://github.com/tile-ai/tilelang/pull/494) ("Introduce quantize components of TileLang", 2025-05-14), and that same commit's body already had `(val >> …) & mask` with no sign-extension — and the correct sign-extending sibling `_tir_packed_int_to_int_to_float` was added right beside it in the same file. So the defect dates to the function's introduction, not a later regression; the only changes since are NFC relocations (the `tvm.tir`→`tirx` rename in #2216 and a type-handling/lint pass in #1448), which reformatted the line but did not alter the logic.
